runhouse/rns/secrets.py

```diff
@@ -146,100 +156,122 @@ def get(
 
-        secrets = cls.load_resp_data(resp).get(provider, {})
+        secrets = cls.load_resp_data(resp).get(provider_name, {})
         if not secrets:
-            raise Exception(f"Failed to load secrets for {provider}")
+            raise Exception(f"Failed to load secrets for {provider_name}")
 
-        cls_name = cls.provider_cls_name(provider)
-        p = cls.get_class_from_name(cls_name)
+        p = cls.builtin_provider_class_from_name(provider_name)
 
         if save_to_env and p is not None:
-            logger.info(f"Saving secrets for {provider} to local config")
+            logger.info(f"Saving secrets for {provider_name} to local config")
             p.save_secrets(secrets)
 
         return secrets
 
     @classmethod
-    def delete(cls, providers: List[str]):
-        """Delete secrets from Vault for the specified providers"""
+    def delete_from_vault(cls, providers: Optional[List[str]] = None):
+        """Delete secrets from Vault for specified providers (builtin or custom).
+        If none are provided, will delete secrets for all providers which have been enabled in the local environment."""
+        providers = providers or cls.builtin_providers(as_str=True)
         for provider in providers:
-            provider_cls_name = cls.provider_cls_name(provider)
-            p = cls.get_class_from_name(provider_cls_name)
-            if p is None:
-                continue
-
-            p.delete_secrets_from_vault()
-            logger.info(f"Successfully deleted {provider} secrets from Vault")
+            url = f"{rns_client.api_server_url}/{cls.USER_ENDPOINT}/{provider}"
+            resp = requests.delete(url, headers=rns_client.request_headers)
+            if resp.status_code != 200:
+                logger.error(
+                    f"Failed to delete secrets from Vault: {json.loads(resp.content)}"
+                )
 
     @classmethod
-    def delete_secrets_from_vault(cls):
-        resp = requests.delete(
-            f"{rns_client.api_server_url}/{cls.USER_ENDPOINT}/{cls.PROVIDER_NAME}",
-            headers=rns_client.request_headers,
-        )
-        if resp.status_code != 200:
-            raise Exception(f"Failed to delete {cls.PROVIDER_NAME} secrets from Vault")
+    def to(cls, hardware: Union[str, "Cluster"], providers: Optional[List] = None):
+        """Copy secrets to the desired hardware for a list of builtin providers. If no providers are specified
+        will load all builtin providers that are already enabled."""
+        if isinstance(hardware, str):
+            from runhouse import cluster
+
+            hardware = cluster(name=hardware)
+
+        hardware_name = hardware.name
+        if not hardware.is_up():
+            raise RuntimeError(
+                f"Hardware {hardware_name} is not up. Run `hardware_obj.up()` to re-up the cluster."
+            )
+
+        provider_secrets: list = cls.load_provider_secrets(providers=providers)
+        if not provider_secrets:
```

Comment:
It looks like we'll only pull from vault if a user passes in a list and none of them are found (vs. if just one or two aren't found locally). Maybe we should check the len of provider_secrets and see if it's shorter than providers (if providers is not None), and if so, try to pull from vault and fill those in? Then, if some still haven't been found, we should raise an exception saying secrets can't be found for x providers.